src/hardware/mouse/mouse_config.cpp

```diff
@@ -0,0 +1,349 @@
+        // issues.
+};
+
+bool MouseConfig::ParseSerialModel(const std::string &model_str,
+                                   MouseModelCOM &model, bool &auto_msm)
+{
+	if (model_str == list_models_com[0]) {
+		model    = MouseModelCOM::Microsoft;
+		auto_msm = false;
+		return true;
+	} else if (model_str == list_models_com[1]) {
+		model    = MouseModelCOM::Logitech;
+		auto_msm = false;
+		return true;
+	} else if (model_str == list_models_com[2]) {
+		model    = MouseModelCOM::Wheel;
+		auto_msm = false;
+		return true;
+	} else if (model_str == list_models_com[3]) {
+		model    = MouseModelCOM::MouseSystems;
+		auto_msm = false;
+		return true;
+	} else if (model_str == list_models_com[4]) {
+		model    = MouseModelCOM::Microsoft;
+		auto_msm = true;
+		return true;
+	} else if (model_str == list_models_com[5]) {
+		model    = MouseModelCOM::Logitech;
+		auto_msm = true;
+		return true;
+	} else if (model_str == list_models_com[6]) {
+		model    = MouseModelCOM::Wheel;
+		auto_msm = true;
+		return true;
+	}
+
+	return false;
+}
+
+const std::vector<uint16_t> &MouseConfig::GetValidMinRateList()
+{
+	return list_rates;
+}
+
+static void config_read(Section *section)
+{
+	assert(section);
+	const Section_prop *conf = dynamic_cast<Section_prop *>(section);
+	assert(conf);
+	if (!conf)
+		return;
+
+	// Settings changeable during runtime
+
+	mouse_config.dos_immediate = conf->Get_bool("dos_mouse_immediate");
+	mouse_config.raw_input     = conf->Get_bool("mouse_raw_input");
+	GFX_SetMouseRawInput(mouse_config.raw_input);
+
+	// Settings below should be read only once
+
+	if (mouse_shared.ready_config_mouse) {
+		MOUSE_NotifyStateChanged();
+		return;
+	}
+
+	// Default mouse sensitivity
+
+	auto sensitivity_from_str = [](const std::string &str) {
+		const int32_t base_value = 50;
+		int32_t ret_val          = base_value; // default
+		bool invalid_input       = false;
+
+		if (str.find('.') != std::string::npos) {
+			// Parameter supplied in a form of floating point
+			float tmp_float = 1.0f;
+			try {
+				tmp_float = std::stof(str);
+			} catch (...) {
+				invalid_input = true;
+				tmp_float     = 1.0f;
+			}
+			// We need 'tmp_float' to be positive for further calculations
+			if (tmp_float < 0.0f) {
+				tmp_float = -tmp_float;
+				ret_val   = -1;
+			} else if (tmp_float > 0.0f)
+				ret_val = 1;
+			else
+				ret_val = 0;
+			// Now calculate user steps as a logarithm
+			if (ret_val != 0) {
+				tmp_float = std::log(tmp_float) / std::log(2.0f);
+				tmp_float *= mouse_predefined.sensitivity_double_steps;
+				tmp_float += static_cast<float>(base_value);
+				const auto tmp_rounded = std::lround(
+				        std::max(tmp_float, 1.0f));
+				ret_val = clamp_to_int8(ret_val * tmp_rounded);
+			}
+		} else if (!str.empty()) {
+			// Parameter supplied in a form of integer
+			try {
+				ret_val = std::stoi(str);
+			} catch (...) {
+				invalid_input = true;
+				ret_val       = base_value;
+			}
+		}
```

Comment:
@FeralChild64 , I forgot that users of the original sensitivity will expect percentages instead of a floating-point scalar (so we should probably keep that interface).

To answer my previous feedback comment:
 - User wants to cut sensitivity into 1/3rd? Use 33
 - User wants to boost sensitivity by 50%? Use 150
 - User wants 8-fold sensitivity? Use 800
 - User wants to inverted behavior? Use -100 (for the Y-axis)

Hopefully it means we can just divide the user's int16_t value by 100.0f to get floating-point scale, and your existing code will apply straight-away.

Very sorry for this churn (but I think it makes sense) - as other conf settings (like volume) are also percentages (and not scalars).

Reply:
To keep it clear - you want me to change the current floating point code to accept percentage? Or to get rid of the exponential system completely? I wanted to have values 0-99 (exponential sensitivity) for:
- easier display in the command line tool (less characters needed)
- easier to create a GUI with a slider; as changing sensitivity from 100% to 150% will bring a very noticeable effect, but change from 500% to 550% won't

I don't know how the volume is implemented in DOSBox Staging, but usually it uses decibels, which are also logarithmic.

Reply:
> you want me to change the current floating point code to accept percentage?

Yes. All DOSBoxes support percentage based mouse sensitivity (and mixer volumes):
 - "10" -> 1/10th of original
 - "50" -> 50% or half of original
 - "200" -> 2-fold more than original
 - "-100" -> invert the axis, at original sensitivity

A number in any form, provided without prefixed units, should be interpreted as this percentage.

For exponential input, I agree with all the benefits, and yes, all DOSBoxes also support decibel volume input when the number is prefixed with a `d`.
 - "d0" -> original/default level
 - "d-3" -> -3 dB is roughly half of original
 - "d3" -> +3 dB is roughly 2-fold more than original

I think a similar exponential prefixed-approach could be used for the mouse sensitivity.

> I wanted to have values 0-99 (exponential sensitivity)

The audio code parses the percent and decibel using the `parse_value` and `parse_prefixed_value` functions. Here's a simple adaptation if we assumed an `x` prefix for the mouse exponential input:

```
#include <string_utils.h>

const std::string arg = // user input;

// did the user provide a non-prefixed percentage?
if (auto v = parse_value(arg, -10000, 10000); v)
  exponential = percent_to_exponential(*val); 
 
// did the user provide an 'x' prefixed exponential?
else if (v = parse_prefixed_value('x', user_input, 0, 99); v)
  exponential = *v;
```